### DISCUSSION TITLE
How does the '--random' work

### DISCUSSION THREAD

**Discussion by chr1shung** [pymodbus-dev/pymodbus#1163] [Help] [answered]
Hi all, 

I'm new to python and modbus and recently I'm using pymodbus REPL server as a simulator of the real modbus device .
Everything works fine I have another service acting as client and it can read the data from pymodbus correctly.
However I'm wondering what does the `--random` option do ? I thought it determined whether to randomize the value to-be-read every `r` reads but I still got the same value if I don't manually update it with another REPL client.

Thanks for the help .

**Comment by janiversen** [comment]
It depends on which version you are using, in 3.0.0 there are a known bug, which are solved on dev.

**Reply by chr1shung** [reply]
I'm actually build the image with the latest dev branch and ran:
```shell
docker run --rm -it -p 5020:5020 --name pymodbus --network-alias pymodbus \
    local/pymodbus pymodbus.server --host pymodbus run --modbus-server tcp --framer socket --unit-id 1 --random 1
```

My client (another dockerized service) is reading on address 8 every five seconds and it always read `0` (I guess is the default value ?)

**Reply by janiversen** [reply]
In this case I need to refer you to @dhoomakethu who is maintaining the repl server.

**Reply by dhoomakethu** [reply]
The random value are generated only for input registers and discrete inputs. If you are reading coils or holding register, you will get the default values

**Reply by chr1shung** [reply]
I see.

Is there any way to know I'm reading input registers and discrete inputs ?
Or it is determined by the function code sent from the client? (it is explicitly reading holding register)

**Reply by dhoomakethu** [reply]
Function codes is what determines it, look in to the helper methods exposed by clients

**Reply by chr1shung** [reply]
Is there any restriction that certain data type is only allowed to reside in certain address location ?

For example I'm able to read input registers in address 8 but unable to read discrete inputs in address 4000:
```
2022-11-07 02:15:02,638 MainThread      ERROR    pdu            :118      Exception Response(130, 2, IllegalAddress)
```

edit: nvm I'm unable to read address 4000 no matter what

**Reply by dhoomakethu** [reply]
Depends on how the datablock is setup , the default data block size is `100` , so if you try to read anything above 100 will result in `IllegalAddress` exception. refer https://github.com/riptideio/pymodbus/blob/ce8be32c2ec9f89d795125393511d6ae10e88816/pymodbus/server/reactive/default_config.py#L39

**Reply by chr1shung** [reply]
Can I clone and update this config and use `--modbus-config` to overwrite default setting ?
Also `max_register_value: 65535`, Does that mean theoretically I can change `block_size` up to 65535?

**Reply by dhoomakethu** [reply]
You can pass a json file with modified values to overwrite the values, the readme has the instructions.

**Reply by serroba** [reply]
I asked a similar in an old PR, but I will ask it again here as this seems related.
> Or it is determined by the function code sent from the client? 

https://en.wikipedia.org/wiki/Modbus#Modbus_object_types

I am just starting to learn about modbus, so here I may be just wrong, but based on what I read there, it seems to me that certain addresses are dedicated to certain object types.

The default config though, seems to always start on the same address location, which is confusing to me. Maybe I am just misunderstandign how to interpret the default JSON file and the modbus specification.

When I use the `pymodbus.console` client, I would have expected to be able to send the following command for example:

```
client.read_holding_registers address 4001 count 10 slave 1
```
 That fails due to the configuration of the default JSON config, but I would like to understand if I am misusing the client/console or am I misunderstanding how the modbus specification works?

**Reply by chr1shung** [reply]
> You can pass a json file with modified values to overwrite the values, the readme has the instructions.

Is it possible to overwrite config.json with environment variable ?

**Reply by dhoomakethu** [reply]
I believe it should be possible. I will check and confirm.

**Reply by dhoomakethu** [reply]
@hahattan its not enabled as of now, pass the custom config file instead . But can be do-able.

**Reply by chr1shung** [reply]
ok, thanks for the clarification.

**Reply by chr1shung** [reply]
I have a use case that data has a change rate percentage every read, for eample change rate 50 means modbus data has 50% possibility to be changed(randomized) on each read.
Do you think it's reasonable to update `--random` to this behavior or add a new flag for it ?
I'm guessing these are the code snippet needs update:
https://github.com/riptideio/pymodbus/blob/b80ce221272c84b090e34663c4b491a53e34701a/pymodbus/server/reactive/main.py#L165-L181
I can help sumbitting PR if needed, thanks

**Reply by dhoomakethu** [reply]
`--random` right now instructs how many reads to wait before the value is randomised.  You will have to tinker this [line](https://github.com/riptideio/pymodbus/blob/b80ce221272c84b090e34663c4b491a53e34701a/pymodbus/server/reactive/main.py#L167) 
```
if not (self._read_counter.get(_block_type) % self._randomize):
```
If you use `--random=1` it is going to change for every read.  May be we will have to introduce another command line option to control the probability of the change . The simplest hack could be to use `--random=2` so that its changed only on every second read.

**Reply by janiversen** [reply]
you could also use the new datastore simulator (on dev, will be part of 3.1), it allow you to write a custom function for each register or group of registers. I use to simulate solar production energy flow during 24 hours compressed into 3600 steps.

**Reply by chr1shung** [reply]
You're right, however in my use case an upstream client is reading all these modbus values and sending to somewhere else **only if their values are changed**. 
The current `--random` leads to that all the values will be sent every 'r' read, and what I want is only a small portion of datas are chagned and being sent everytime I read.

I'll fork the repo and try to update it ony my side, thanks for all your help!

Edit:  just saw @janiversen's comment about datastore simulator, def gonna check it and see if it helps. Thanks again

**Reply by janiversen** [reply]
please submit your work as a pull request as other might have the same needs.

**Reply by dhoomakethu** [reply]
Thats a good idea what @janiversen suggested, I will spend sometime if I can use the datastore simulator in `repl` server as well.

**Reply by dhoomakethu** [reply]
but if you beat me to it with your PR, we are happy to accept it.

**Reply by janiversen** [reply]
@dhoomakethu I am already working on a similar thing: integrating the repl server and simulator into the web server so you can remotely control the server.

**Reply by janiversen** [reply]
@dhoomakethu but I am focusing on the web server first, so if you integrate the simulator datastore I can use that.

**Reply by dhoomakethu** [reply]
Ok, I am yet to go through the new simulator datastore, Will spend some this week to understand it and try to add it to the REPL server. With that said, I will also check if I can use the latest and greatest features in pymodbus and update this simulator if possible. https://github.com/riptideio/modbus-simulator

**Comment by janiversen** [comment, marked as answer]
Please check that your dev contains commit https://github.com/riptideio/pymodbus/commit/30f71669353114b8a00e5bf60be2e9ebccc65515
It was added November 2, so very new.

**Reply by chr1shung** [reply]
and btw, does the docker image still public accessible ?
```shell
➜  ~ docker pull riptideio/pymodbus
Using default tag: latest
Error response from daemon: pull access denied for riptideio/pymodbus, repository does not exist or may require 'docker login': denied: requested access to the resource is denied

```

**Comment by janiversen** [comment]
There is an error in the documentation. After the documentation update was merged, it turned out that the riptideio organization is inaccessible for this and a couple of other urgent problems so the image could not be stored there.

To help the contributor doing docker I used my private account “janiversen” where it is publicly available, sorry about the confusion.

**Reply by chr1shung** [reply]
Is it `janiversen/pymodbus` ?
Tried pull it and get the same error.

**Reply by janiversen** [reply]
Seems there was an upload error, so it is not available until the contributor who made it repairs it. The pymodbus team do not maintain docker and other packer images.

**Reply by janiversen** [reply]
see #1178

**Reply by chr1shung** [reply]
I see, I will build the image by myself.
Thanks for the clarification

**Reply by janiversen** [reply]
If you want to maintain it, we are happy to include it (again) in our build process, just submit a pull request.
